Rename test and add sutiable checks and exceptions

amit-jain opened the pull request:
## Description

- Rename ReplicationIT -> PublishEndToEndIT
- Remove #activateAndDelete test
- Add Replication pre-checks and also check replication queue to check after test
- Add Service Up/Down pre-checks
- Introduce appropriate Exceptions and error codes to help to classify the failures if any

## Related Issue

https://github.com/adobe/aem-test-samples/issues/32

## Motivation and Context

The current ReplicationIT is an actual end to end test but the name suggests that the responsibility of the failure is for 1 component only. But that is not true as the assertions on presence/absence of a page for ADD/DELETE case relies on the page being accessible through the dispatcher and publish.

## How Has This Been Tested?

Changes tested on a cloud environment

## Screenshots (if appropriate):

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

Commit:
* Rename from ReplicationIT -> PublishEndToEndIT
* Remove testActivateAndDelete as redundant

Commit:
Check replication status on author
- Check replication queue unblocked along with agents availability
- proper replication response and replication exceptions

Commit:
Check Author/Publish health before tests
- Currently only logs the response from the health check error

Commit:
Throw relevant exceptions on page checks if not successful after activation or deactivation
- Use code 403 to also throw exceptions on the first check before retry

Commit:
Disable redirect and throw exception if redirect code returned

Commit:
Document different exceptions and error code

Commit:
Only log and not throw error in case of redirects
Logging to  delienate activation and deactivation from rest

amit-jain commented:
@dulvac @tmaret Appreciate if you could review the changes.

The comments and changes from the previous [PR](https://github.com/adobe/aem-test-samples/pull/37) have been largely incorporated. 
I have classified exceptions into major areas and added appropriate ERRORCODE to disinguish different failures within those.

Commit:
trivial doc updates

tmaret commented on smoke/README.md:
```diff
@@ -6,3 +6,20 @@ This is a collection of generic smoke tests to validate the basic functionality
+| ReplicationException | QUEUE_BLOCKED               | Replication queue blocked before test                              |
+|                      | ACTIVATION_REQUEST_FAILED   | Replication action failed                                          |
+|                      | DEACTIVATION_REQUEST_FAILED | Deactivation action failed                                         |
+|                      | ITEM_NOT_REPLICATED         | Item to be replicated still available on agent queues              |
```
Suggest to `s/ITEM_NOT_REPLICATED/ACTION_NOT_REPLICATED` and `Replicated action still pending in agent queues` to align with the terminology already used in other reasons

tmaret commented on smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/replication/ReplicationClient.java:
```diff
@@ -0,0 +1,221 @@
+        return this.doPost("/bin/replicate.json", formEntityBuilder.build(), Collections.emptyList());
+    }
+
+    private SlingHttpResponse activateInternal(String agent, String nodePath) throws ClientException {
```
Could be merged with `deactivateInternal` method above to avoid duplicating the code.

tmaret commented on smoke/README.md:
```diff
@@ -6,3 +6,20 @@ This is a collection of generic smoke tests to validate the basic functionality
+|                      | ACTIVATION_REQUEST_FAILED   | Replication action failed                                          |
+|                      | DEACTIVATION_REQUEST_FAILED | Deactivation action failed                                         |
+|                      | ITEM_NOT_REPLICATED         | Item to be replicated still available on agent queues              |
+|                      | REPLICATION_UNAVAILABLE     | Replication agents not available before test                       |
```
Suggest to `s/REPLICATION_UNAVAILABLE/REPLICATION_NOT_AVAILABLE` to align with the other error code.

tmaret commented on smoke/README.md:
```diff
@@ -6,3 +6,20 @@ This is a collection of generic smoke tests to validate the basic functionality
+|                      | DEACTIVATION_REQUEST_FAILED | Deactivation action failed                                         |
+|                      | ITEM_NOT_REPLICATED         | Item to be replicated still available on agent queues              |
+|                      | REPLICATION_UNAVAILABLE     | Replication agents not available before test                       |
+| ServiceException     | AUTHOR_DOWN                 | Author is down before test                                         |
```
Suggest to `s/DOWN/NOT_AVAILABLE` because it can be that the instance is up but not reachable.

tmaret commented on smoke/README.md:
```diff
@@ -6,3 +6,20 @@ This is a collection of generic smoke tests to validate the basic functionality
+|                      | ITEM_NOT_REPLICATED         | Item to be replicated still available on agent queues              |
+|                      | REPLICATION_UNAVAILABLE     | Replication agents not available before test                       |
+| ServiceException     | AUTHOR_DOWN                 | Author is down before test                                         |
+|                      | PUBLISH_DOWN                | Publish is down before test                                        |
```
As above

tmaret reviewed (approved):
LGTM with suggested minor changes.

Commit:
Review changes
- Change error codes
-- ITEM_NOT_REPLICATED -> ACTION_NOT_REPLICATED
-- REPLICATION_UNAVAILABLE -> REPLICATION_NOT_AVAILABLE
-- AUTHOR_DOWN -> AUTHOR_NOT_AVAILABLE
-- PUBLISH_DOWN -> PUBLISH_NOT_AVAILABLE
- remove deactivateInternal

amit-jain commented:
Incorporated suggested changes from review

amit-jain commented:
@andreituicu @dulvac Gentle reminder to review the PR please. Thanks!!

amit-jain commented:
@dulvac @andreituicu @volteanu Gentle reminder for review or please tag others which can help in review and merge of this PR.
Thanks!!

dulvac commented on smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:
```diff
@@ -0,0 +1,97 @@
+    public ContentPublishRule
+        contentPublishRule = new ContentPublishRule(root, cqBaseClassRule.authorRule, cqBaseClassRule.publishRule);
+
+	/**
```
```suggestion
        /**
```

dulvac commented on smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java:
```diff
@@ -0,0 +1,299 @@
+            res  = getPublishClient().doStreamRequest(request, null);
+            
+            // Special handling for 401,403, logging for 301,302
+            if (null != res && (res.getStatusLine().getStatusCode() == SC_UNAUTHORIZED
```
what happens with the test execution if res is `null`?
You could check for null and fail and you wouldn't have to check for null in all the `if`s

dulvac reviewed (approved):
small couple of comments, but it looks ok.
Thanks!

dulvac committed:
Update smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
